pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidProof,
+        ErrorValidatingXCM,
+        ErrorDeliveringXCM,
+        /// An order for this para id already exists
+        OrderAlreadyExists,
+        /// The para id is not a parathread
+        NotAParathread,
+        /// There are too many in-flight orders, buying cores will not work until some of those
+        /// orders finish.
+        InFlightLimitReached,
+        /// There are no collators assigned to this parathread, so no point in buying a core
+        NoAssignedCollators,
+        /// This collator is not assigned to this parathread
+        CollatorNotAssigned,
+        /// The `XcmWeights` storage has not been set. This must have been set by root with the
+        /// value of the relay chain xcm call weight and extrinsic weight
+        XcmWeightStorageNotSet,
+    }
+
+    /// Proof that I am a collator, assigned to a para_id, and I can buy a core for that para_id
+    #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
+    #[scale_info(skip_type_params(T))]
+    pub struct BuyCoreCollatorProof<T: Config> {
+        account: T::AccountId,
+        // TODO
+        _signature: (),
+    }
+
+    /// Set of parathreads that have already sent an XCM message to buy a core recently.
+    /// Used to avoid 2 collators buying a core at the same time, because it is only possible to buy
+    /// 1 core in 1 relay block.
+    #[pallet::storage]
+    //pub type InFlightOrders<T: Config> = StorageMap<_, Blake2_128Concat, ParaId, (), OptionQuery>;
+    pub type InFlightOrders<T: Config> =
+        StorageValue<_, BoundedBTreeSet<ParaId, T::MaxParathreads>, ValueQuery>;
+
+    /// This must be set by root with the value of the relay chain xcm call weight and extrinsic
+    /// weight limit. This is a storage item because relay chain weights can change, so we need to
+    /// be able to adjust them without doing a runtime upgrade.
+    #[pallet::storage]
+    pub type XcmWeights<T: Config> = StorageValue<_, XcmWeightsTy<T>, OptionQuery>;
+
+    #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
+    #[scale_info(skip_type_params(T))]
+    pub struct XcmWeightsTy<T> {
+        pub buy_execution_cost: u128,
+        pub weight_at_most: Weight,
+        pub _phantom: PhantomData<T>,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Buy a core for this parathread id.
+        /// Collators should call this to indicate that they intend to produce a block, but they
+        /// cannot do it because this para id has no available cores.
+        /// The purchase is automatic using XCM, and collators do not need to do anything.
+        // Note that the collators that will be calling this function are parathread collators, not
+        // tanssi collators. So we cannot force them to provide a complex proof, e.g. against relay
+        // state.
+        #[pallet::call_index(0)]
+        // TODO: weight
```

Comment:
```suggestion
```

Reply:
Actually this is still using the wrong weight, because I expect it to be more expensive than `force_buy_core` because of the proof validation. But proof validation is not being implemented in this PR, so it is not possible to write a benchmark for this.